### ISSUE TITLE
Redesign / simplify element creation DSL

### ISSUE THREAD

**Issue by Exerosis** [kwebio/kweb-core#168]
Right now things like render are extensions on ElementCreator<*>. Obviously, we need that scope to do anything but the syntax isn't the nicest. It would instead be possible to have syntax like:
```kotlin
data.each {
    li().text(it.something)
}
```
You can achieve this by adding some extensions to ElementCreator itself just like this:
```kotlin
interface ElementCreator {
    ...
    inline operator fun <First, Return> 
        (ElementCreator.(First) -> (Return))
        .invoke(first: First) = 
        invoke(this@ElementCreator, first)
    inline operator fun <First, Second, Return> 
        (ElementCreator.(First, Second) -> (Return))
        .invoke(first: First, second: Second) = 
        invoke(this@ElementCreator, first, second)

    //additional specializations as needed.
    ...
}
```kotlin
then you can make each like:
```kotlin
val <Type> KVal<List<Type>>.each //perhaps you want some typealiasing lol.
get(): ElementCreator.((Type) -> (Unit)) -> (Return) = { block ->
    render(this@each) { it.forEach(block) }
}
```
It certainly isn't perfect, but it's the closest you can get to a double extension in Kotlin (as far as I know).
The problems start showing up when you need to "overload" these kinds of methods... obviously vals can't be overloaded so the only thing you can do is put the overloads in their own files and then if you need both import with aliases.
This can also be used to avoid having .new {} and some other things.

Also perhaps check this out:
https://gist.github.com/Exerosis/ef32fb37525c852c56373eacbd71ea85
You might consider adopting a similar container system in place of something like shoebox and KVal/KVar. (the Toggled aspect might not be important, it exists for powerful scoping and state-to-state side effects.)
Also, check out the way that I handle mapping of mutables :P
Also, it's plenty easy to create database/file based containers I even have a temporary redis one for testing: https://srcb.in/h3WrPGicxT

I also recommend you make KVal a delegate provider and I even more highly recommend you do not support
element.text = someKVar maybe this is just me... but it feels a bit unclear to me.

Lastly, out of curiosity why do you have .text() and .new rather than just having:
```kotlin
body {
    div {
        h1 { "some header" }
    }
    blah blah
    p { "some text" }
}
```
Either that or having text be a method param.
p(text = "some text")
h1("some text")
and reserving {} for children.

Sorry for writing such a book, really cool library I think this kind of thing is the future for sure. (also I do not intend to be condescending, I just really enjoy thinking about architecture and design. ~ Exerosis

**Comment by sanity** [comment]
Appreciate your thoughts, feedback on API ergonomics is particularly helpful because I've been using the current API for so long it's easy to become blind to its flaws.

Will give it some thought before responding in detail.

**Comment by Exerosis** [comment]
If you are interested we could also work together on some of these ideas and then perhaps the successful ones can be brought into the fold. Also have you ever considered using the KotlinJS compiler to transform the DSL content into JS. It might be possible to have the logic both client and server-side and then have hybrid client-server rendering.

**Comment by frnknglrt** [comment]
Hello,

I found the same issue with using KVal and lists. When rendering full pages with multiple kvars the code becomes deeply nested and the readability of the code is decreased.

I ended up wrapping the affected ElementCreator<*>.someElement functions in a small ‘useState’ extension that wraps all the kvars. This is quite similar to how react hooks are functioning.

Small example:

**Comment by sanity** [comment]
Just a quick note to let you know I haven't forgotten about this, have been stuck into another project and it's been difficult to come up for air.  This is high on my todo list and will get to it soon :)

**Comment by sanity** [comment]
I've been playing around with this. It's fairly easy to replace:

```kotlin
p().new {
  table().new {
    ...
  }
}
```

With:

```kotlin
p {
  table {
      ...
   }
}
```
See [here](https://github.com/kwebio/kweb-core/blob/549122df9fac9fd158de90afed086beab1010c87/src/main/kotlin/kweb/prelude.kt#L32) for an example of the implementation.

As you point out, the issue occurs when the element has attributes, in which case we're back to:

```kotlin
h1(mapOf("class", "heading")) {
  ...
}
```
The fundamental reason to have `text()` and `new { ... }` is that setting an element's text is similar but different to creating child elements, which is different in turn to other modifications that can be made to an element such as setting an `<input>`s `value`.

I'm definitely open to the possibility that there is a better syntax for this, but it will need to account for some of these subtleties.

**Comment by sanity** [comment]
Doing a little more playing around:

```kotlin
    Kweb(port = 1243) {
        doc.body {
            ul()

            ul {
                li()
            }

            ul (mapOf("class" to "dog")) {
                li()
            }
        }
    }
```

**Comment by sanity** [comment]
Implemented in [0.8.0 release](https://github.com/kwebio/kweb-core/releases/tag/0.8.0).
